text/0000-exception-safe-event-handlers.md

```diff
@@ -0,0 +1,164 @@
+
+```rust
+    let mut this = cx.this();
+    let cb = cx.argument::<JsFunction>(0)?;
+    let handler = EventHandler::new(cb);
+    // or:      = EventHandler::bind(this, cb);
+    thread::spawn(move || {
+        for i in 0..100 {
+            // do some work ....
+            thread::sleep(Duration::from_millis(40));
+            // schedule a call into javascript
+            handler.schedule(move |cx| {
+                // successful result to be passed to the event handler
+                Ok(cx.number(i).upcast())
+            }
+        }
+    });
+```
+
+Here the `EventHandler` "captures" `this` and `cb` and calls the closure from the JavaScript thread with the context (`cx`). The successful result value produced by the closure is then passed as the second argument of the event handler callback, with `null` passed as the first argument, indicating no error occurred.
+
+*Note:* The closure is sent to the main JavaScript thread so every captured value will be moved.   
+
+This approach should be familiar to Rust programmers as it is the same as `std::thread::spawn` uses.
+
+## Low-level API
+
+For cases where you need more control, Neon offers a lower-level primitive, the `schedule_with` method. This method also takes an event launcher, but Neon does not automatically call the event handler. Instead, the event launcher receives a `neon::Context`, the `this` context, and the event handler function object, and is given total control over whether and how to call the event handler.
+
+If the event launcher throws a JavaScript exception, it behaves like an uncaught exception in the Node event loop.
+
+Example for providing the current progress of a background operation:
+
+```rust
+    let mut this = cx.this();
+    let cb = cx.argument::<JsFunction>(0)?;
+    let handler = EventHandler::new(cb);
+    // or:      = EventHandler::bind(this, cb);
+    thread::spawn(move || {
+        for i in 0..100 {
+            // do some work ....
+            thread::sleep(Duration::from_millis(40));
+            // schedule a call into javascript
+            handler.schedule_with(move |cx, this, cb| {
+                // call the event handler callback
+                let args = vec![
+                    cx.null().upcast(),
+                    cx.number(i).upcast()
+                ];
+                cb.call(args)?;
+                Ok(())
+            }
+        }
+    });
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+```rust
+struct EventHandler {
+
+    pub fn new<'a, C: Context<'a>, T: Value>(cx: &C, this: Handle<T>, callback: Handle<JsFunction>) -> Self;
+
+    pub fn schedule<T, F>(&self, arg_cb: F)
+        where T: Value,
+              F: for<'a> FnOnce(&mut EventContext<'a>) -> JsResult<'a, T>,
+              F: Send + 'static;
+
+    pub fn schedule_with<F>(&self, arg_cb: F)
+        where F: FnOnce(&mut EventContext, Handle<JsValue>, Handle<JsFunction>) -> NeonResult<()>,
```

Comment:
Can we make `JsValue` generic so that users don't need to `upcast`?

Reply:
The user takes the `JsValue` as an argument, so they don't upcast it. They do have to downcast it if they know it's a particular type but that's not tracked statically. To track it statically we'd need to make `EventHandler` generic in the type of the `this`-binding, maybe something like:

```rust
struct EventHandler<This=JsObject> {
    pub fn new<'a, C: Context<'a>>(cx: &C, this: Handle<This>, callback: Handle<JsFunction<This>>) -> Self;

    pub fn schedule<T, F>(&self, arg_cb: F)
        where T: Value,
              F: for<'a> FnOnce(&mut EventContext<'a>) -> JsResult<'a, T>,
              F: Send + 'static;

    pub fn schedule_with<F>(&self, arg_cb: F)
        where F: FnOnce(&mut EventContext, Handle<This>, Handle<JsFunction<This>>) -> NeonResult<()>,
              F: Send + 'static;

}
```

Reply:
Oh, you probably meant for `schedule()` rather than `schedule_with()`!

Reply:
Wait, I'm still confused 😂  It looks like `schedule()` already _is_ generic and didn't need the upcast in the example!